### PR TITLE
Compute tria intersections

### PR DESCRIPTION
**Pull request by fdrmrc**
- Fixed numbering of vertices in rectangle-rectangle intersection 0fbcacd13f20c660dc1f885d66783f0f289eea6d

 - Added new function `intersect_and_get_quads()`. It returns a ```std::vector<
  std::tuple<typename dealii::Triangulation<dim0, spacedim>::active_cell_iterator,typename dealii::Triangulation<dim1, spacedim>::active_cell_iterator,Quadrature<spacedim>>>```

- Extended namespace `dealii::NonMatching` so that it contains `compute_intersection` (that works for cells only) and the new functions for triangulations

### CODE REVIEW

File: include/compute_intersections_of_tria.h [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    dealii::Quadrature<spacedim>>>
+  intersect_and_get_quads(
+    const GridTools::Cache<dim0, spacedim> &space_cache,
+    const GridTools::Cache<dim1, spacedim> &immersed_cache,
+    const unsigned int                      degree);
+
+}
+
+
+
```
Comment (luca-heltai) [inline include/compute_intersections_of_tria.h:53]:
```suggestion
```

File: include/compute_intersections_of_tria.h [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+#include <deal.II/grid/grid_tools.h>
+#include <deal.II/grid/grid_tools_cache.h>
+
+#include <set>
+#include <tuple>
+#include <vector>
+
+#include "compute_intersection_of_cells.h"
+
+
```
Comment (luca-heltai) [inline include/compute_intersections_of_tria.h:36]:
```suggestion
```

File: include/compute_intersections_of_tria.h [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+namespace dealii::NonMatching
+{
+  template <int dim0, int dim1, int spacedim>
+  std::vector<std::tuple<
+    typename dealii::Triangulation<dim0, spacedim>::active_cell_iterator,
+    typename dealii::Triangulation<dim1, spacedim>::active_cell_iterator,
+    dealii::Quadrature<spacedim>>>
+  intersect_and_get_quads(
```
Comment (luca-heltai) [inline include/compute_intersections_of_tria.h:45]: I would rather call also this function `compute_intersection`, leveraging C++ overloading capabilities.

Can you add a documentation for the function, and keep both functions in the same header file?

**Review by luca-heltai — CHANGES_REQUESTED** [review]

**Comment by fdrmrc** [comment]
Thanks Luca for the review. I added operator overloading and documentation for `compute_intersection`. I renamed the files and fixed the #includes in the `gtests` folder.

**Comment by luca-heltai** [comment]
Documentation is still missing. Would you mind adding some documentation? Remember: someone who has never seen this before, should understand what it does, and how to use it. :)

File: source/compute_intersections.cc [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+   * @tparam dim0
+   * @tparam dim1
+   * @tparam spacedim
+   * @param cell0 A `cell_iteratator` to the first cell
+   * @param cell1 A `cell_iteratator` to the first cell
+   * @param degree The degree of the `Quadrature` you want to build there
+   * @param mapping0 The `Mapping` object describing the first cell
+   * @param mapping1 The `Mapping` object describing the second cell
+   * @return Quadrature<spacedim>
+   */
```
Comment (luca-heltai) [inline source/compute_intersections.cc:110]: This should go in the header file, not in the source file. :)

File: source/compute_intersections.cc [hunk 5]
```diff
@@ -0,0 +1,387 @@
...
+   * @tparam dim1 Intrinsic dimension of the ambient grid
+   * @tparam spacedim
+   * @param space_cache
+   * @param immersed_cache
+   * @param degree Degree of the desired quadrature formula
+   * @return std::vector<std::tuple<
+   * typename dealii::Triangulation<dim0, spacedim>::active_cell_iterator,
+   * typename dealii::Triangulation<dim1, spacedim>::active_cell_iterator,
+   * dealii::Quadrature<spacedim>>>
+   */
```
Comment (luca-heltai) [inline source/compute_intersections.cc:233]: Same here. The documentation goes together with the declaration, not in the cc file, but in the h file.

**Review by luca-heltai — CHANGES_REQUESTED** [review]

**Comment by luca-heltai** [comment]
If you need help with the rebase, let me know, and I'll do it for you (or explain to you how to do it, so you know it next time).

File: source/compute_intersections.cc [hunk 6]
```diff
@@ -0,0 +1,375 @@
...
+                                                           test_intersection));
+              }
+          }
+      }
+
+    return cells_with_quads;
+  }
+
+} // namespace dealii::NonMatching
+
```
Comment (luca-heltai) [inline source/compute_intersections.cc:287]: Add an `#else`  and `#endif` here, where you have the default implementations that simply throws exceptions saying that you need CGAL, and then instantiate all of the functions irrespective of wether CGAL exists or not. This makes sure you don't get errors in user codes at link time, but just at run time.

**Review by luca-heltai — CHANGES_REQUESTED** [review]

**Comment by fdrmrc** [comment]
- I added here `create_coupling_sparsity_pattern_with_exact_intersections()` . As you suggested, this depends on the output of `compute_intersection()`, which is a vector of tuples.
- `compute_intersection` has been improved by avoiding to store the intersected cells in a `std::set`, so that only a double for over the trees is needed. 

This sparsity pattern has been checked in `reduced_lagrange.cc`, and runs fine

File: source/compute_intersections.cc [hunk 7]
```diff
@@ -0,0 +1,372 @@
...
+
+
+#else
+
+Assert(DEAL_II_WITH_CGAL,
+       ExcMessage("This function needs CGAL to be installed, "
+                  "but cmake could not find it."));
```
Comment (luca-heltai) [inline source/compute_intersections.cc:284]:
```suggestion
template <int dim0, int dim1, int spacedim>
  std::vector<std::tuple<
    typename dealii::Triangulation<dim0, spacedim>::active_cell_iterator,
    typename dealii::Triangulation<dim1, spacedim>::active_cell_iterator,
    dealii::Quadrature<spacedim>>>
  compute_intersection(const GridTools::Cache<dim0, spacedim> &,
                       const GridTools::Cache<dim1, spacedim> &,
                       const unsigned int                     )
  {
Assert(false,
       ExcMessage("This function needs CGAL to be installed, "
                  "but cmake could not find it."));
                  return {};
}
     
                 template <int dim0, int dim1, int spacedim>
  dealii::Quadrature<spacedim>
  compute_intersection(
    const typename dealii::Triangulation<dim0, spacedim>::cell_iterator &,
    const typename dealii::Triangulation<dim1, spacedim>::cell_iterator &,
    const unsigned int                                                  ,
    const dealii::Mapping<dim0, spacedim> &,
    const dealii::Mapping<dim1, spacedim> &)
  {
               Assert(false,
       ExcMessage("This function needs CGAL to be installed, "
                  "but cmake could not find it."));
                  return {};
                  }
```

**Pull request merged**